### DISCUSSION TITLE
응집도와 결합도에 대해서 설명해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#46] [OOP]
.

**Comment by kunsanglee** [comment]
응집도(Cohesion)는 하나의 모듈이나 클래스 내부의 요소들이 서로 얼마나 밀접하게 관련되어 있는지를 나타내는 척도입니다. 응집도가 높은 모듈은 특정한 작업이나 목적을 위해 서로 잘 연결되어 있으며, 이는 코드의 재사용성, 이해하기 쉽고, 유지보수가 쉬운 코드로 이어집니다.

결합도(Coupling)는 서로 다른 모듈이나 클래스 간의 의존성 정도를 나타냅니다. 결합도가 낮다는 것은 모듈 간의 상호 의존성이 적다는 것을 의미하며, 이는 각 모듈이나 클래스가 독립적으로 작동할 수 있음을 뜻합니다. 결합도가 낮을수록 시스템의 유연성과 확장성이 증가하며, 변경 사항이 한 부분에 국한되어 다른 부분에 미치는 영향을 최소화할 수 있습니다.

응집도와 결합도는 서로 반비례하는 경향이 있습니다. 즉, 시스템을 설계할 때 응집도를 높이려고 노력하면 자연스럽게 결합도가 낮아지는 경향이 있습니다. 효과적인 소프트웨어 설계는 모듈이나 클래스가 단일한 목적 또는 기능에 집중하게 하여 응집도를 높이고, 동시에 각 모듈이나 클래스가 서로 독립적으로 작동할 수 있도록 하여 결합도를 낮게 유지하는 것을 목표로 합니다.

**Comment by seunghye218** [comment]
응집도는 단일 모듈이나 클래스 내 요소의 내부 구조와 관련성에 중점을 두는 반면, 결합도는 서로 다른 클래스 간의 상호 의존성에 중점을 둡니다.

응집도는 단일 모듈이나 객체 내의 책임이 서로 연관되어 함께 작동하여 특정 목표나 기능을 달성하는 정도를 나타냅니다.  
응집도가 높은 객체는 목적이 명확하고 잘 정의되어 있어 이해, 유지 관리 및 재사용이 더 쉽습니다.

결합도가 높은 클래스는 서로 의존도가 높기 때문에 변경이나 교체가 더 어렵고 의도하지 않은 부작용이 발생할 위험이 높아집니다.

응집력이 높고 느슨하게 결합된 객체는 이해, 테스트, 리팩터링이 더 쉽습니다. 시스템의 한 부분이 변경되더라도 다른 부분에서 의도하지 않은 결과가 발생할 가능성이 적기 때문입니다.

**Comment by le2sky** [comment]
응집도는 모듈에 포함된 내부 요소들이 연관되어 있는 정도를 나타냅니다.
결합도는 의존성의 정도를 나타내며, 다른 모듈에 대해 얼마나 많은 지식을 갖고 있는지를 나타냅니다.
일반적으로 좋은 설계란 높은 응집도와 낮은 결합도를 가진 모듈로 구성된 설계를 의미합니다.
응집도와 결합도는 변경과 관련이 깊습니다. 또한, 캡슐화의 정도가 응집도와 결합도에 영향을 미칩니다.
캡슐화를 지키면, 모듈 안의 응집도는 높아지고 모듈 사이의 결합도는 낮아집니다.
따라서, 응집도와 결합도를 고려하기 전에 먼저 캡슐화를 향상하기 위해 노력해야 합니다.

**Comment by GIVEN53** [comment]
응집도 -> 내부 요소간 연관 정도
결합도 -> 의존성 정도
좋은 설계 -> 응집도 높음 결합도 낮음
유연성, 확장성, 유지보수, 테스트
캡슐화
